src/java_bytecode/java_bytecode_language.cpp

```diff
@@ -340,32 +346,62 @@ static void gather_needed_globals(
 static void gather_field_types(
   const typet &class_type,
   const namespacet &ns,
+  const class_hierarchyt &ch,
   ci_lazy_methodst &lazy_methods)
 {
   const auto &underlying_type=to_struct_type(ns.follow(class_type));
   for(const auto &field : underlying_type.components())
   {
     if(field.type().id()==ID_struct || field.type().id()==ID_symbol)
-      gather_field_types(field.type(), ns, lazy_methods);
+      gather_field_types(field.type(), ns, ch, lazy_methods);
     else if(field.type().id()==ID_pointer)
     {
       // Skip array primitive pointers, for example:
       if(field.type().subtype().id()!=ID_symbol)
         continue;
-      const auto &field_classid=
-        to_symbol_type(field.type().subtype()).get_identifier();
-      if(lazy_methods.add_needed_class(field_classid))
-        gather_field_types(field.type().subtype(), ns, lazy_methods);
+      initialize_needed_classes_from_pointer(
+        to_pointer_type(field.type()), ns, ch, lazy_methods);
     }
   }
 }
 
-/// See output
-/// \par parameters: `entry_points`: list of fully-qualified function names that
+/// Build up list of methods for types for a specific pointer type. See
+/// `initialize_needed_classes` for more details.
+/// \param pointer_type: The type to gather methods for.
+/// \param ns: global namespace
+/// \param ch: global class hierarchy
+/// \param [out] lazy_methods: Populated with all Java reference types whose
+///   references may be passed, directly or indirectly, to any of the functions
+///   in `entry_points
+static void initialize_needed_classes_from_pointer(
+  const pointer_typet &pointer_type,
+  const namespacet &ns,
+  const class_hierarchyt &ch,
+  ci_lazy_methodst &lazy_methods)
+{
+  const symbol_typet &class_type=to_symbol_type(pointer_type.subtype());
+  const auto &param_classid=class_type.get_identifier();
+  std::vector<irep_idt> class_and_parents=
+    ch.get_parents_trans(param_classid);
+
+  class_and_parents.push_back(param_classid);
```

Comment:
I would prefer to write `lazy_methods.add_needed_class(param_classid)` here instead. At the moment, calling `push_back` here may cause a reallocation which can be avoided.